### PR TITLE
Addon-docs: Support docs-only MDX && update docs mode nav UI

### PR DESCRIPTION
**Pull request by shilman**
Issue: #6644

## What I did

- [x] Updated MDX compiler to gen a "docsOnly" story when MDX has no stories
- [x] Updated UI to show "docsOnly" stories in both canvas & docs modes
- [x] Updated `DOCS_MODE` UI to show only component-level nodes
- [x] Added example to `official-storybook`
- [x] Documented future enhancements in #7304

Here's how a docs-only node shows up in the UI:

Storybook

Here's how the tree looks in `--docs` mode:

Storybook

## How to test

```
cd examples/official-storybook
yarn storybook
yarn storybook --docs
```

### CODE REVIEW

**Comment by shilman** [comment]
Enhancements: #7304

File: lib/ui/src/containers/nav.js [hunk 1]
```diff
@@ -95,6 +96,64 @@ const createMenu = memoize(1)((api, shortcutKeys, isFullscreen, showPanel, showN
...
+  return filtered;
+};
+
+export const removeLeafStories = stories => {
+  const filtered = cloneDeep(stories);
+  Object.values(stories).forEach(story => {
+    const { id, isLeaf, parent } = story;
+    if (isLeaf) {
+      if (!filtered[parent].isLeaf) {
+        filtered[parent].isLeaf = true;
```
Comment (tmeasday) [inline lib/ui/src/containers/nav.js]: This logic seems wrong. It is possible to have both leaves and non leaves as children. For instance

```
GroupA
|-- GroupB
|      \-- Story1
\-- Story 2
```

I think there are some more complex examples in the code base, eg. in https://github.com/storybookjs/storybook/blob/next/lib/ui/src/components/sidebar/treeview/treeview.mockdata.js#L1.

I think to be correct, you'd need to delete `story` from `parent.children`, *then* set it to `isLeaf` if it is empty

Comment (tmeasday) [reply lib/ui/src/containers/nav.js]: Can we just do this (without clone) in a fairly simple way:

```js
return stories.filter(s => !s.isLeaf).map(({ children, ...story }) => {
  const newChildren = children.filter(s => !s.isLeaf);
  return newChildren.length === 0 ? {...story, isLeaf: true } : {...story, children: newChildren};
});
```

File: lib/ui/src/containers/nav.js [hunk 2]
```diff
@@ -95,6 +96,64 @@ const createMenu = memoize(1)((api, shortcutKeys, isFullscreen, showPanel, showN
...
+  if (node.parent) {
+    const parent = stories[node.parent];
+    parent.children = parent.children.filter(child => child !== id);
+    if (!parent.children.length) {
+      recursiveRemove(stories, parent.id);
+    }
+  }
+};
+
+export const removeDocsOnlyStories = stories => {
```
Comment (tmeasday) [inline lib/ui/src/containers/nav.js]: What is this for? It only seems to be used in tests?

File: lib/ui/src/containers/nav.js [hunk 3]
```diff
@@ -95,6 +96,64 @@ const createMenu = memoize(1)((api, shortcutKeys, isFullscreen, showPanel, showN
...
+  return filtered;
+};
+
+export const flattenDocsOnlyStories = stories => {
+  const filtered = cloneDeep(stories);
+  Object.values(stories).forEach(story => {
+    const { id, isLeaf, parent, parameters } = story;
+    if (isLeaf && parameters && parameters.docsOnly) {
+      if (!filtered[parent].isLeaf) {
+        filtered[parent].isLeaf = true;
```
Comment (tmeasday) [inline lib/ui/src/containers/nav.js]: In theory this has the same problem as `removeLeafStories`.

Also you could probably do a no-`cloneDeep` version very similar to below

**Review by tmeasday — CHANGES_REQUESTED** [review]
I think there is a bug here

File: lib/ui/src/containers/nav.js (hunk 3 above)
Comment (tmeasday) [reply lib/ui/src/containers/nav.js]: In fact a `removeLeaves(stories, predicate?)` function would make sense and remove the need for the duplicate logic.

**Pull request closed without merging**